pm: esp32c6: Power management support

raffarost opened the pull request:
Adds power management (light and deep sleep) support to ESP32C6

Commit:
boards: esp32c6_devkitc: Test files for power management

Config files for pm testing

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
pm: clock_control: esp32c6: Peripherals clock init review

Clock is disabled at startup for LP core peripherals, greatly
improving power consumption in deep sleep mode.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

raffarost commented on dts/riscv/espressif/esp32c6/esp32c6_common.dtsi:
```diff
@@ -66,14 +83,13 @@
 		};
 
-		rtc: rtc@600b000 {
+		rtc: rtc@600b0000 {
```
I believe this address is currently not used, so it was a non-functional fix

raffarost commented:
Can we have some help to review and (eventually) approve this?

kartben commented on samples/boards/esp32/deep_sleep/src/main.c:
the `light_sleep` sample should also probably be updated?

raffarost replied:
As all ESP32 use the same wake-up source for testing (GPIO), the light sleep sample is generic for all chips.

kartben commented on samples/boards/esp32/deep_sleep/src/main.c:
```diff
@@ -81,6 +82,12 @@ int main(void)
 			ESP_EXT1_WAKEUP_ANY_HIGH);
+
+	/* enable pull-down on ext1 pins to avoid random wake-ups */
+	rtc_gpio_pullup_dis(EXT_WAKEUP_PIN_1);
+	rtc_gpio_pulldown_en(EXT_WAKEUP_PIN_1);
+	rtc_gpio_pullup_dis(EXT_WAKEUP_PIN_2);
+	rtc_gpio_pulldown_en(EXT_WAKEUP_PIN_2);
```
does this mean the kconfig doc for `EXAMPLE_EXT1_WAKEUP` should be updated? right now it is talking about what might happens with floating pins

kartben reviewed (dismissed):
Some hopefully helpful feedback from a quick test.
I tested on m5stack nano c6, but didn't have time to pull both HAL changes for the USB serial fix and this one so I don't have the console logs handy. 
Device seems to properly go to deep sleep, down to 17uA power consumption, and apparently starts to wake up from RTC alarm but fails -- this is what I see on the `west espressif monitor` 
console, ad infinitum

```
SHA-256 comparison failed:
Calculated: 6eceed53b5cc6323d01c6037b3e2844681be2b96f4a7657320673bd6fa1913cb
Expected: 0000000020530000000000000000000000000000000000000000000000000000
Attempting to boot anyway...
entry 0x408024f0
0x408024f0: __start at C:/Users/kartben/zephyrproject/zephyr/soc/espressif/common/loader.c:229

To exit from IDF monitor please use "Ctrl+]" or "Ctrl+C"
ESP-ROM:esp32c6-20220919
Build:Sep 19 2022
rst:0x7 (TG0_WDT_HPSYS),boot:0xc (SPI_FAST_FLASH_BOOT)
Saved PC:0x408020ee
0x408020ee: esp_rom_regi2c_write_mask at C:/Users/kartben/zephyrproject/modules/hal/espressif/components/esp_rom/patches/esp_rom_regi2c_esp32c6.c:182 (discriminator 1)

SPIWP:0xee
mode:DIO, clock div:2
load:0x40800000,len:0x9418
0x40800000: _esp32c6_vector_table at ??:?

load:0x40809430,len:0x1884
0x40809430: ?? at C:/Users/kartben/zephyrproject/modules/hal/espressif/components/driver/gpio/gpio.c:74
```

I have increased `CONFIG_ESP_SLEEP_WAIT_FLASH_READY_EXTRA_DELAY` to 5000 (and even 50000!) us to no avail.

EXT1 wakeup doesn't seem to work at all as nothing happens when I connect "G2" pin of my M5Stack to 5v.

Thanks!

Commit:
samples: pm: esp32: Enable pull-down for simple testing

Sample was modified to enable pull-down when using EXT1 pins for
wake-up on deep sleep sample in order to simplify testing in case
user doesn't notice external pull-down will be needed otherwise.

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

Commit:
pm: esp32c6: Power management support

Power management support (light/deep sleep) for ESP32C6

Signed-off-by: Raffael Rostagno <raffael.rostagno@espressif.com>

raffarost commented:
> @raffarost your latest push fixes the RTC wake-up on my nanoC6 :)

Thank you for spotting this and letting me know! :100: 
I have no idea how I didn't see that behaviour in the testing I did, as I even had the current figure in my mind.
With esp32c6_devkitc (which has a jumper for this purpose) I measured 7.6 uA in _deep sleep_. Datasheet mentions 7 uA as minimum, so it is kind of close. I'll see if I can chip off some more uA's in the future.

The reason I asked for help approving this is because we made some updates on HAL side and one file is missing (which is in this PR) and causes other PR's CI to fail if HAL version is advanced. I apologize I still make mistakes and am likely unable to coordinate all things the rightful way.

kartben commented:
> Edit: there is an Impostor SHA warning here, so I also need to find the proper way to fix it.

It sounds to me like a bug in the newly added check. cc @carlescufi

kartben commented:
> EXT1 wakeup doesn't seem to work at all as nothing happens when I connect "G2" pin of my M5Stack to 5v.

Well, if I actually connect to G2 it works 🤦🏼  I was connecting to G1... :)

raffarost commented:
> > Edit: there is an Impostor SHA warning here, so I also need to find the proper way to fix it.
> 
> It sounds to me like a bug in the newly added check. cc @carlescufi

I think the problem is I advanced SHA to a newer commit in HAL (regarding USB JTAG clock control) to be able to include the clock optimization in this PR, as it made sense to release the optimization along the PM support. I guess that for fixing this I'll need either to remove the commit regarding clock optimization and create a separate PR or create a PR to update HAL SHA and possibly the Kconfig file regarding delay a power up delay setting, which is in this PR. Again, my apologies.

marekmatej commented on soc/espressif/Kconfig:
```diff
@@ -56,4 +56,37 @@ config RTC_CLK_CAL_CYCLES
 	    If the crystal could not start, it will be switched to internal RC.
 endmenu
 
+menu "Sleep Config"
```
Can we remove this fix from this PR, please? There is a draft in https://github.com/zephyrproject-rtos/hal_espressif/pull/311

marekmatej commented on drivers/clock_control/clock_control_esp32.c:
```diff
@@ -41,6 +41,9 @@
+#include <hal/usb_serial_jtag_ll.h>
+#include <esp_private/esp_pmu.h>
+#include <ocode_init.h>
 #endif /* CONFIG_SOC_SERIES_ESP32xx */
```
compliance check usually dont like if #endif comments differs from the #if statement

marekmatej commented on drivers/clock_control/clock_control_esp32.c:
```diff
@@ -655,11 +675,18 @@ static int clock_control_esp32_init(const struct device *dev)
 
 	rst_reas = esp_rom_get_reset_reason(0);
+
+#if defined(CONFIG_SOC_SERIES_ESP32C6)
```
this if/endif structure here is starting to be a mess. Could we make it more readable (and maintainable)?

marekmatej commented on west.yml:
```diff
@@ -157,7 +157,7 @@ manifest:
         - hal
     - name: hal_espressif
-      revision: 87e7902d7184a8280b4d13bce79801a723f4ddd8
+      revision: ee0a24b31f4559ce499b89c74155b3692b406b9c
```
please update the hash

sylvioalves commented on samples/boards/esp32/deep_sleep/Kconfig:
```diff
@@ -7,17 +7,18 @@ config EXAMPLE_EXT1_WAKEUP
-	  When triggering a wake up, connect one or both of the pins to
-	  HIGH. Note that floating pins may trigger a wake up.
+	  This option enables wake-up from deep sleep using GPIO2 and
+	  GPIO4. The sample enables internal pull-downs on EXT1 pins to
```
`pull-downs` -> `pull-down`

kartben reviewed (commented):
+1
fwiw the clang-format warning is legit IMO so should there be another round of updates, it likely makes sense to address it

kartben commented:
Yay!
